crates/sui-replay-2/src/artifacts.rs

```diff
@@ -0,0 +1,264 @@
+impl<'b> ArtifactManager<'b> {
+    /// Creates a new `ArtifactManager` with the given base path and whether overrides are allowed.
+    pub fn new(base_path: &'b Path, overrides_allowed: bool) -> anyhow::Result<Self> {
+        std::fs::create_dir_all(base_path).map_err(|e| {
+            anyhow::anyhow!(
+                "Failed to create base path for replay artifacts at {}: {e}",
+                base_path.display()
+            )
+        })?;
+        Ok(ArtifactManager {
+            base_path,
+            overrides_allowed,
+        })
+    }
+
+    /// Create an `ArtifactManager` for this artifact type, rooted at the given `base_path`.
+    pub fn member<'a>(&'a self, artifact: Artifact) -> ArtifactMember<'a, 'b> {
+        ArtifactMember {
+            artifact_path: self.base_path.join(artifact.as_file()),
+            manager: self,
+            artifact_type: artifact,
+        }
+    }
+}
+
+impl Artifact {
+    /// Returns the string representation of the artifact type.
+    pub const fn as_str(&self) -> &str {
+        match self {
+            Artifact::Trace => "trace",
+            Artifact::TransactionEffects => "transaction_effects",
+            Artifact::ForkedTransactionEffects => "forked_transaction_effects",
+            Artifact::TransactionGasReport => "transaction_gas_report",
+        }
+    }
+
+    /// Encoding type for each artifact.
+    pub fn encoding_type(&self) -> EncodingType {
+        match self {
+            Artifact::Trace => EncodingType::JsonCompressed,
+            Artifact::ForkedTransactionEffects
+            | Artifact::TransactionEffects
+            | Artifact::TransactionGasReport => EncodingType::Json,
+        }
+    }
+
+    /// Returns the file for the artifact, including its encoding type. The returned `PathBuf` is
+    /// not rooted.
+    pub fn as_file(&self) -> PathBuf {
+        PathBuf::from(format!("{}.{}", self.as_str(), self.encoding_type().ext()))
+    }
+}
+
+/// Deserialization methods for `ArtifactManager`.
+impl ArtifactMember<'_, '_> {
+    pub fn exists(&self) -> bool {
+        self.artifact_path.exists()
+    }
+
+    /// Deserialize the artifact into json. This should always succeed if the artifact exists and
+    /// is well-formed.
+    pub fn get_json(&self) -> anyhow::Result<serde_json::Value> {
+        let file_content = std::fs::read(&self.artifact_path)?;
+        let contents = match self.artifact_type.encoding_type() {
+            EncodingType::Json => file_content,
+            EncodingType::JsonCompressed => {
+                let mut buf = Vec::new();
+                zstd::bulk::decompress_to_buffer(&file_content, &mut buf)?;
+                buf
+            }
+        };
+        Ok(serde_json::from_slice(&contents)?)
+    }
+
+    /// Try to get the trace reader if the artifact type is a trace.
+    /// If the artifact type is not `Trace` `None` is returned.
+    pub fn try_get_trace(&self) -> Option<anyhow::Result<MoveTraceReader<std::fs::File>>> {
```

Comment:
Where is this function used?

Reply:
Nowhere yet! This is just to get the basic APIs there.